libyul/optimiser/StructuralSimplifier.cpp

```diff
@@ -0,0 +1,137 @@
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <libyul/optimiser/StructuralSimplifier.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/optimiser/Utilities.h>
+#include <libyul/AsmData.h>
+#include <libdevcore/CommonData.h>
+#include <libdevcore/Visitor.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+namespace {
+
+ExpressionStatement makePopExpressionStatement(langutil::SourceLocation const& _location, Expression _expression)
+{
+	return {_location, FunctionalInstruction{
+		_location,
+		solidity::Instruction::POP,
+		{std::move(_expression)}
+	}};
+}
+
+}
+
+void StructuralSimplifier::operator()(Block& _block)
+{
+	pushScope(false);
+	simplify(_block.statements);
+	popScope();
+}
+
+void StructuralSimplifier::simplify(std::vector<yul::Statement> &_statements)
+{
+	GenericFallbackReturnsVisitor<boost::optional<vector<Statement>>, If, Switch, ForLoop> const visitor(
+		[&](If& _ifStmt) -> boost::optional<vector<Statement>> {
+			if (_ifStmt.body.statements.empty())
+				return {{makePopExpressionStatement(_ifStmt.location, std::move(*_ifStmt.condition))}};
+			if (expressionAlwaysTrue(*_ifStmt.condition))
+				return {std::move(_ifStmt.body.statements)};
+			else if (expressionAlwaysFalse(*_ifStmt.condition))
+				return {{}};
+			return {};
+		},
+		[](Switch& _switchStmt) -> boost::optional<vector<Statement>> {
+			if (_switchStmt.cases.size() == 1)
+			{
+				auto& switchCase = _switchStmt.cases.front();
+				auto loc = locationOf(*_switchStmt.expression);
+				if (switchCase.value)
+					return {{If{
+						std::move(_switchStmt.location),
+						make_shared<Expression>(FunctionalInstruction{
+								std::move(loc),
+								solidity::Instruction::EQ,
+								{std::move(*switchCase.value), std::move(*_switchStmt.expression)}
+						}), std::move(switchCase.body)}}};
+				else
+					return {{
+						makePopExpressionStatement(loc, std::move(*_switchStmt.expression)),
+						std::move(switchCase.body)
+					}};
+			}
+			else
+				return {};
+		},
+		[&](ForLoop& _forLoop) -> boost::optional<vector<Statement>> {
+			if (expressionAlwaysFalse(*_forLoop.condition))
+				return {std::move(_forLoop.pre.statements)};
+			else
+				return {};
+		}
+	);
+
+	iterateReplacing(
+		_statements,
+		[&](Statement& _stmt) -> boost::optional<vector<Statement>>
+		{
+			visit(_stmt);
+			auto result = boost::apply_visitor(visitor, _stmt);
+			if (result.is_initialized())
```

Comment:
`is_initialized` seems to be deprecated. Why not just use `if (result)`? And furthermore, `result` could profit from an explicit type :)

Reply:
If that is too verbose, you could use `using OptionalStatements = boost::optional<vector<Statement>>` at the beginning of the function.

Reply:
Haha, I had ``if (result)``, but changed to ``if (result.is_initialized())`` for "explicitness" - but I didn't know that it's deprecated, I'll change it. And yeah, I can write down the type.